### DISCUSSION TITLE
How to put barriers between dispatch() calls?

### DISCUSSION THREAD

**Discussion by kjpgit** [gpuweb/gpuweb#4434] [Q&A] [answered]
Hi, I read the webgpu spec "synchronization" section again, but it didn't answer my question.

This is my code:

            computePass.setPipeline(bin_pipeline);
            computePass.setBindGroup(0, bin_bg)
            computePass.dispatchWorkgroups(Math.ceil(scene.num_shapes() / constants.WG_BIN_WORKLOAD))

            computePass.setPipeline(bin2_pipeline);
            computePass.setBindGroup(0, bin2_bg)
            computePass.dispatchWorkgroups(Math.ceil(constants.NUM_TILES_TOTAL / constants.WG_BIN2_WORKLOAD))

            computePass.setPipeline(bin3_pipeline);
            computePass.setBindGroup(0, bin3_bg)
            computePass.dispatchWorkgroups(Math.ceil(scene.num_shapes() / constants.WG_BIN_WORKLOAD))
            
            
This is my PIX output:
![image](https://github.com/gpuweb/gpuweb/assets/8010221/9dab0b4b-b8bc-48da-9f68-57a613371955)

Questions:

1) Are these 3 dispatch calls intended to be separated by an implicit barrier, so they don't run concurrently?  

2) I can't tell from the PIX output if they are running concurrently.  I do NOT want them running concurrently -- they have read/write acccess to the same buffers in their BGs, but I need the 3 dispatches to run serially.

3) Why is a computePipeline called a "pipeline" when there's only one shader in it?  The terminology is quite strange.  It seems like the object that *contains* these pipelines is actually the pipeline.

**Comment by greggman** [comment]
> 1. Are these 3 dispatch calls intended to be separated by an implicit barrier, so they don't run concurrently?

They won't run concurrently if they depend on the same resources. WebGPU handles this for you. There is no need to manually synchronize.

**Reply by kjpgit** [reply]
Based on your response, I still don't have a clear mental model.

If two shaders only READ the same buffers, they won't run in parallel?   What is the purpose of setting storage modes like read/read_write in your bind groups, then?  Typically, memory models allow concurrency for read-only data, because there can't be races.

**Reply by Kangz** [reply]
The WebGPU spec says that dispatches are each their own synchronization scope so in terms of memory it is as-if they were run serially. However implementation should optimize this so that two dispatches that only read from the same resource (but don't write to any shared resource) run in parallel. Dawn, Chromium's implementation, should already be doing this.

**Reply by kjpgit** [reply, marked as answer]
So basically if I compare a WebGPU buffer with a typical SQL Database row:
- a buffer/row can be locked multiple times for read (shared lock) by concurrent dispatches/transactions
- a buffer/row can be locked exclusively for write, and can not have any read locks at the same time
- dispatches/transactions appear to have been run serially, even though gpu/RDBMS can choose to run in parallel.

**Reply by Kangz** [reply]
That comparison looks correct!
